### ISSUE TITLE
Runtime and expanded nodes against NA*

### ISSUE THREAD

**Issue by DiTo97** [archettialberto/neural_weighted_a_star#1]
Hi @archettialberto,

I am quite interested in your work, especially in the introduction of the trade-off parameter $\epsilon$. As the title suggests, I would like to know if you have conducted explicit experiments on the runtimes of your model against the baselines, mainly NA*. Of course, the runtime of forward passes is highly correlated with the number of expanded nodes (which you report in Table 3 in the paper), but if explicit traces of the runtimes are available, I would be interested to know them.

About the expanded nodes, I noticed that they are generally higher than NA*'s (even for $\epsilon = 14$), despite achieving a (much) better cost ratio overall. Do you know how large $\epsilon$ should be in order to reach a number of expanded nodes comparable to NA*? If not, what's your intuition behind this gap with NA*'s search space shrinkage capabilities?

Finally, have you conducted any ablation studies on the learned heuristic function, maybe swapping it out for the fixed admissible Chebyshev heuristic (Eq. 2 in the paper)?

**Comment by archettialberto** [comment]
Hi @DiTo97!

Thanks for your interest in our work. Unfortunately, I did not keep track of the explicit runtimes of the algorithms. However, from what I recall from the experiments, I can say that
* training **BBA*** (and consequently **NWA***) is slower than **NA***. In the former case, during backpropagation, several instances of A* must be solved on CPU. Despite heavy parallelization with [ray](https://github.com/ray-project/ray), it still takes more time than **NA***, which executes entirely on GPU. However, the time difference, despite being noticeable, is not by orders of magnitude. As a rule of thumb, the training time of **NWA*** is comparable to **BBA*** when $\epsilon = 0$ and to **NA*** when $\epsilon \gg 0$;
* evaluating **BBA***, **NWA***, and **NA*** takes roughly the same time, since no backward pass is needed. A slight advantage for **NA*** persists, but the difference is much less noticeable, once the networks are trained.

Concerning the second question, my intuition about **NA*** being so good at expanding few nodes comes from the fact that its heuristic is quite "cheaty": in the original **NA*** [paper](https://arxiv.org/abs/2009.07476), $H$ (fixed by design and not admissible) is much greater than $W$, resulting in node expansions that are extremely unlikely to deviate from straight-line paths. In fact, **NA*** has no theoretical guarantees about the resulting path optimality. It just happens that in most of the samples in the datasets, paths are generally straight from the source to the target.
I think that **NWA*** could in principle reach the number of expanded nodes of **NA***, but only for values of $\epsilon$ that are extremely high. In such a case, the theoretical guarantees of **NWA*** would be useless in practice (that's why experiments with very high $\epsilon$ are not included in the paper). If your target application is just concerned with efficiency, I would suggest you to stick with **NA***, which trains faster ;)

I'm not sure if I understood correctly your question, but **NWA*** running with $H_C$ (Eq. 2 of the paper) is equivalent to **NWA*** with $\epsilon = 0$, which can be found in the results reported.

**Comment by DiTo97** [comment]
Thank you @archettialberto for the in-depth explanation of all the topics I brought up.

You're absolutely right about $H_c$ and $\epsilon = 0$. I overlooked their equivalence.

I am closing the issue!
